File: frontend/src/stores/aiDetection.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { NovelAPI, type AIDetectionResponse } from '@/api/novel'

type DetectionStatus = 'idle' | 'running' | 'success' | 'error'

export interface DetectionState extends AIDetectionResponse {
  status: DetectionStatus
  error_message?: string
  updatedAt?: number
  content_hash?: string
}

const defaultState: DetectionState = {
  status: 'idle',
  confidence: undefined,
  available_uses: undefined,
  segments: [],
  text_hash: undefined,
  error_message: undefined,
  updatedAt: undefined,
  content_hash: undefined
}

export const useAIDetectionStore = defineStore('ai-detection', () => {
  const detections = ref<Record<string, DetectionState>>({})

  const keyOf = (projectId: string, chapterNumber: number) => `${projectId}:${chapterNumber}`

  function getState(projectId: string, chapterNumber: number): DetectionState {
    const key = keyOf(projectId, chapterNumber)
    return detections.value[key] || { ...defaultState }
  }

  function setState(projectId: string, chapterNumber: number, state: DetectionState) {
    const key = keyOf(projectId, chapterNumber)
    detections.value[key] = { ...state, updatedAt: Date.now() }
  }

  function invalidate(projectId: string, chapterNumber: number) {
    const key = keyOf(projectId, chapterNumber)
    detections.value[key] = { ...defaultState }
  }

  async function fetchLatest(projectId: string, chapterNumber: number) {
    const result = await NovelAPI.getLatestAIDetection(projectId, chapterNumber)
    setState(projectId, chapterNumber, { ...result, status: result.status || 'success' })
    return getState(projectId, chapterNumber)
  }

  async function runDetection(projectId: string, chapterNumber: number, text?: string, timeoutSeconds?: number) {
    const previous = getState(projectId, chapterNumber)
    setState(projectId, chapterNumber, { ...previous, status: 'running', error_message: undefined })
    try {
      const result = await NovelAPI.detectChapterAI(projectId, chapterNumber, text, timeoutSeconds)
      setState(projectId, chapterNumber, { ...result, status: result.status || 'success', content_hash: result.content_hash })
      return getState(projectId, chapterNumber)
    } catch (error) {
      const message = error instanceof Error ? error.message : '检测失败'
      setState(projectId, chapterNumber, { ...previous, status: 'error', error_message: message })
      throw error
    }
  }

  return {
    getState,
    setState,
    runDetection,
    invalidate,
    fetchLatest,
  }
})

File: backend/app/schemas/ai_detection.py
from typing import List, Optional

from pydantic import BaseModel, Field


class AIDetectionRequest(BaseModel):
    text: Optional[str] = Field(default=None, description="待检测文本，可为空表示使用当前章节选定版本内容")
    timeout_seconds: Optional[float] = Field(
        default=None,
        ge=1.0,
        description="检测超时时长（秒），为空则使用系统默认"
    )


class SegmentLabel(BaseModel):
    label: int = Field(description="0=人工,1=AI,2=疑似")
    text: str = Field(description="对应片段文本")


class AIDetectionResponse(BaseModel):
    status: str = Field(description="idle|running|success|error")
    confidence: Optional[float] = Field(default=None, description="整体 AI 浓度置信度，0-1")
    available_uses: Optional[int] = Field(default=None, description="剩余可用检测次数")
    segments: List[SegmentLabel] = Field(default_factory=list, description="分段检测结果")
    text_hash: Optional[str] = Field(default=None, description="文本哈希，便于前端校验结果是否过期")
    error_message: Optional[str] = Field(default=None, description="错误原因，status=error 时返回")
    content_hash: Optional[str] = Field(default=None, description="清洗后的正文哈希，用于前端比对")
